torch/_inductor/codegen/cpp.py

```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
                 self._set_dependent_itervars(args[0])
             if any(arg.is_vec for arg in args if isinstance(arg, CppCSEVariable)):
                 self.is_vec = True
-        # NOTE [dtype of CppCSEVariable]
-        # Deciding dtype according to the current optimization context is not
-        # always accurate since the dtypes are initialized during dtype propagation
-        # at the beginning of the codegen. It is possible that some ops are invoked
-        # during the codegen of the current op and take different dtypes from the
-        # current op.
-        # TODO(jgong5): A more accurate way of deciding the dtype of the variables is to
-        # propagate the dtypes here inside `update_on_args`.
+        # NOTE [Deduce dtype of CppCSEVariable at runtime]
         if (
-            hasattr(V.interpreter, "current_node")
-            and get_current_node_opt_ctx() is not None
-        ):
+            output_dtype := _deduce_output_dtype_by_name(
+                name,
+                *args,
+                **kwargs,
+            )
+        ) is not None:
+            self.dtype = output_dtype
+        elif name == "masked":
+            assert (
+                hasattr(V.interpreter, "current_node")
+                and V.interpreter.current_node.target.startswith("masked_subblock")
+                and get_current_node_opt_ctx() is not None
+            )
             self.dtype = get_current_node_opt_ctx().dtype
```

Comment:
Perhaps just set the dtype per the current node context regardless of whether it is `masked` or not?

Reply:
Perhaps we can also deduce the `masked` data type by inputs' cse variable of other. I have added a TODO here, let's check it later.